server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java

```diff
@@ -791,6 +798,50 @@ public void run() {
       }
     }, 0, 5000, TimeUnit.MILLISECONDS);
 
+    // Periodically check that metadata of tablets matches what is held in memory
+    ThreadPools.createGeneralScheduledExecutorService(aconf).scheduleWithFixedDelay(() -> {
+      final SortedMap<KeyExtent,Tablet> onlineTabletsSnapshot = onlineTablets.snapshot();
+
+      final SortedSet<KeyExtent> userTablets = new TreeSet<>();
+      final SortedSet<KeyExtent> nonUserTablets = new TreeSet<>();
+
+      // Create subsets of tablets based on DataLevel: one set who's DataLevel is USER and another
+      // containing the remaining tablets (those who's DataLevel is ROOT or METADATA).
+      // This needs to happen so we can use .readTablets() on the DataLevel.USER tablets in order
+      // to reduce RPCs.
+      // TODO: Push this partitioning, based on DataLevel, to ample.
+      onlineTabletsSnapshot.forEach((k, v) -> {
+        if (Ample.DataLevel.of(k.tableId()) == Ample.DataLevel.USER) {
+          userTablets.add(k);
+        } else {
+          nonUserTablets.add(k);
+        }
+      });
+
+      List<TabletMetadata> tmdList;
+
+      // gather metadata for all tablets with DataLevel.USER using readTablets()
+      try (TabletsMetadata tabletsMetadata = getContext().getAmple().readTablets()
+          .forTablets(userTablets).fetch(FILES, LOGS, ECOMP, PREV_ROW).build()) {
+        tmdList = new ArrayList<>(IteratorUtils.toList(tabletsMetadata.iterator()));
+      }
+
+      // gather metadata for all tablets with DataLevel.ROOT or METADATA using readTablet()
+      nonUserTablets.forEach(extent -> {
+        TabletMetadata tabletMetadata =
+            getContext().getAmple().readTablet(extent, FILES, LOGS, ECOMP, PREV_ROW);
+        tmdList.add(tabletMetadata);
+      });
+
+      // for each tablet, compare its metadata to what is held in memory
+      for (TabletMetadata tabletMetadata : tmdList) {
```

Comment:
If for some reason a tablet was not found in the metadata table, that could be a problem that this loop would not detect because its only look at what was returned. Maybe the code should loop over the onlineTabletsSnapshot set and look for the corresponding data found from metadata table, logging an error if not found.

Reply:
Do you think it would suffice to check that each KeyExtent in onlineTabletsSnapshot is present within one of the returned TabletMetadata? I'm not too sure what other data you might be referring to. @keith-turner

Reply:
> Do you think it would suffice to check that each KeyExtent in onlineTabletsSnapshot is present within one of the returned TabletMetadata?

That is the gist of what I was thinking about.   Thinking about it a bit more, maybe we don't want to worry about this case.  Under normal conditions a tablet could be deleted after we have grabbed the snapshot of online tablets.  So if we did bother detecting it, we would not want to take any action that may cause alarm.  So if we did detect would only want to log a debug or something.  If we are not going to worry about it or attempt to detect it, then looping over tmdList as you have is fine.  If there is an extent missing from tmdList that is in the snapshot, nothing will be done.